opengovsg/postmangovsg
# feat: Redaction reminder digest

## Problem

Closes #812 

## Solution

**Features**:
- Implemented lambda that will send weekly digest to users to remind them of campaigns with delivery report downloads that will be expiring in the next week.
- When the reminder digest will be sent and which campaign will be included in the digest is controlled by the following two parameters:
    - **Retention period**: How long will we will keep the campaign messages before redacting
    - **Notice period**: Minimum number of days before that we will inform users of an upcoming redaction

## Considerations
### Example notification window
Given the following parameters:
- Retention period: 14 days
- Notice period: 7 days

And considering the following campaigns:
- Campaign 1
    - Sent at: 08/11/2020
    - Expiry date: 22/11/2020
- Campaign 2
    - Sent at: 09/11/2020
    - Expiry date: 23/11/2020
- Campaign 3
    - Sent at: 10/11/2020
    - Expiry date: 24/11/2020

If the digest lambda is triggered on 16/11/2020, only **Campaign 2** and **Campaign 3** will be included in the digest. This is because the digest triggered on 16/11/2020 will only notify users of campaigns that will be expiring between 23/11/2020 00:00 and 30/11/2020 00:00 (inclusive).

**Campaign 1** would have already been included in the digest triggered on 09/11/2020.

_November 2020 calendar for reference:_

<img src='https://user-images.githubusercontent.com/3666479/99347165-54263c80-28d1-11eb-9f5b-85e62d2ab7b1.png' width=250/>

### Minimum 14 days retention period
If we want to give users at least 1 week advance notice of an upcoming redaction, we will need to set retention period to be at least 14 days. 

Given the following parameters:
- Retention period: 7 days
- Notice period: 7 days

And considering the following campaign:
- Campaign 1
    - Sent at: 10/11/2020
    - Expiry date: 17/11/2020

If the digest was sent on the 09/11/2020, with a notification window of 16/11 - 23/11, **Campaign 1** will not appear on any digest before it gets redacted as it was created after the reminder digest for the next week was already sent.

### Retries
Automated retries for this lambda is set to 0. In the event that an error occur while sending out the reminder emails, the following should be done to retry sending:
1. Check logs for list of failed recipients
2. Decide which of the failed recipients should be retried
3. Manually trigger the lambda with the following event payload (refer to AWS console test event `ManualSelectedRecipients`):
```json
{
    "selectedRecipients": ["test@example.com", "test2@example.com"]
}
```

## Tests

**Testing locally**
- Create and send three campaigns `test-1`, `test-2` and `test-3`. 
- Manually modify the `updated_at` value in the `job_queue` table in the following manner:
   - `test-1` - Set value to 8 days ago
   - `test-2` - Set value to 7 days ago
   - `test-3` - Set value to 6 days ago
- Run `redaction-digest` lambda
- You should receive an digest email that include reminders for campaigns `test-2` and `test-3` only since both this campaigns will expire in the notification window in the following week. 

**Testing on staging**
- Same test can be conducted on staging
- You may want to limit the list of recipients to only yourself to avoid spamming other users on staging. This can be achieved by setting your email in the `selectedRecipients` array in the event payload
 
## Deploy Notes

**New Travis environment variables**:
_Note: These variables have already been populated on Travis_
- Production
	- `PRODUCTION_REDACTION_ROLE`
	- `PRODUCTION_REDACTION_DIGEST_DB_URI`
	- `PRODUCTION_REDACTION_DIGEST_DB_READ_REPLICA_URI`
	- `PRODUCTION_REDACTION_DIGEST_SES_FROM`
	- `PRODUCTION_REDACTION_DIGEST_SES_HOST`
	- `PRODUCTION_REDACTION_DIGEST_SES_PASS`
	- `PRODUCTION_REDACTION_DIGEST_SES_USER`
	- `PRODUCTION_REDACTION_DIGEST_SES_PORT` 
	- `PRODUCTION_REDACTION_CRONITOR`
- Staging
	- `STAGING_REDACTION_ROLE`
	- `STAGING_REDACTION_DIGEST_DB_URI`
	- `STAGING_REDACTION_DIGEST_DB_READ_REPLICA_URI`
	- `STAGING_REDACTION_DIGEST_SES_FROM`
	- `STAGING_REDACTION_DIGEST_SES_HOST`
	- `STAGING_REDACTION_DIGEST_SES_PASS`
	- `STAGING_REDACTION_DIGEST_SES_USER`
	- `STAGING_REDACTION_DIGEST_SES_PORT`
	- `STAGING_REDACTION_CRONITOR`

**Deployment**
- Create new database user on production
```sql
CREATE USER serverless_redaction_iam;
GRANT rds_iam TO serverless_redaction_iam;
GRANT SELECT ON campaigns TO serverless_redaction_iam;
GRANT SELECT ON job_queue TO serverless_redaction_iam;
GRANT SELECT ON statistics TO serverless_redaction_iam;
GRANT SELECT ON users TO serverless_redaction_iam;
```
- Enable CloudWatch schedule `redaction-digest-schedule-production`
- Enable Cronitor alerting

## TODOs
- [x] Add manual event trigger for retrying selected recipients
- [x] CloudWatch scheduled events (`redaction-digest-schedule-production` and `redaction-digest-schedule-staging`)
- [x] Test on staging
- [x] Set up Cronitor for staging and production
- [x] Add Cronitor codes to environment variable
